Xamarin.Forms.Core/OnIdiom.cs

```diff
@@ -1,33 +1,116 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xamarin.Forms.Xaml;
+
 namespace Xamarin.Forms
 {
+	[ContentProperty("Idioms")]
 	public class OnIdiom<T>
 	{
-		public T Phone { get; set; }
+		readonly Dictionary<TargetIdiom, T> _legacyValues = new Dictionary<TargetIdiom, T>();
+
+		bool _hasDefault;		
 
-		public T Tablet { get; set; }
+		T _default;
 		
-		public T Desktop { get; set; }
 
-		public T TV { get; set; }
+		public OnIdiom()
+		{
+			Idioms = new List<OnIdiomSetting>();
+		}
+
+		public T Phone
+		{
+			get => GetLegacyValue(TargetIdiom.Phone);
+			set => SetLegacyValue(TargetIdiom.Phone, value);
+		}
+
+		public T Tablet
+		{
+			get => GetLegacyValue(TargetIdiom.Tablet);
+			set => SetLegacyValue(TargetIdiom.Tablet, value);
+		}
+
+		public T Desktop
+		{
+			get => GetLegacyValue(TargetIdiom.Desktop);
+			set => SetLegacyValue(TargetIdiom.Desktop, value);
+		}
+
+		public T TV
+		{
+			get => GetLegacyValue(TargetIdiom.TV);
+			set => SetLegacyValue(TargetIdiom.TV, value);
+		}
 
-		public T Watch { get; set; }
+		public T Watch
+		{
+			get => GetLegacyValue(TargetIdiom.Watch);
+			set => SetLegacyValue(TargetIdiom.Watch, value);
+		}
+
+		public T Default
+		{
+			get { return _default; }
+			set
+			{
+				_hasDefault = true;
+				_default = value;
+			}
+		}
+
+		public IList<OnIdiomSetting> Idioms { get; private set; }
+
+#pragma warning disable RECS0108 // Warns about static fields in generic types
+		static readonly IValueConverterProvider s_valueConverter = DependencyService.Get<IValueConverterProvider>();
+#pragma warning restore RECS0108 // Warns about static fields in generic types
 
 		public static implicit operator T(OnIdiom<T> onIdiom)
 		{
-			switch (Device.Idiom)
+			var currentIdiom = Device.Idiom.ToString("G");
+			foreach (var idiomSetting in onIdiom.Idioms)
 			{
-				default:
-				case TargetIdiom.Phone:
-					return onIdiom.Phone;
-				case TargetIdiom.Tablet:
-					return onIdiom.Tablet;
-				case TargetIdiom.Desktop:
-					return onIdiom.Desktop;
-				case TargetIdiom.TV:
-					return onIdiom.TV;
-				case TargetIdiom.Watch:
-					return onIdiom.Watch;
+				if (idiomSetting.Idiom?.FirstOrDefault(idiom =>
```

Comment:
```suggestion
				if (idiomSetting.Idiom?.Find(idiom =>
```

Reply:
Is there a reason for not using `FirstOrDefault`? The type is `IList` which does not have `Find` method, so the only option would be then to manually write a `foreach` search.

Reply:
I rebased too :-)

Reply:
`FirstOrDefault` and `foreach` are far less performant than `Find` or `for`, especially on Android. We avoid Linq as much as possible because of the enumerator performance hit.